### ISSUE TITLE
Return known addresses and decoded data when available

### ISSUE THREAD

**Issue by sche** [safe-global/safe-client-gateway#840]
**Is your feature request related to a problem? Please describe.**
When getting decoded data in [reponse](https://safe.global/safe-client-gateway/docs/routes/contracts/routes/fn.post_data_decoder.html) we would like to get the known address.

**Provide references to the feature you are implementing that requires this change**
[Problem statement](https://github.com/safe-global/safe-pm/issues/76)
[Mobile issue](https://github.com/safe-global/safe-ios/issues/1710)
[Slack thread](https://gnosisinc.slack.com/archives/CQB729LLF/p1648559663170409)

**Describe the solution you'd like**
The nesting should be capped to be at most 1 level deep

**Describe alternatives you've considered**
Nope

**Additional context**
We want to display decoded data for incoming WalletConnect transactions.

**How it looks currently**

![IMG_0861](https://user-images.githubusercontent.com/2284385/170063363-35732443-a022-436e-b55b-d6c9d1a478ff.jpeg)

**What we aim**

![Screenshot 2022-05-24 at 11 31 23](https://user-images.githubusercontent.com/2284385/170063449-43b582d8-92df-4588-b827-c0d4fb38cbbf.jpeg)

![Screenshot 2022-05-24 at 11 32 10](https://user-images.githubusercontent.com/2284385/170063463-f1d66ab3-8298-4fc3-b15d-869de289539e.jpeg)

**Solution ideas**

Provide `addressInfoIndex` object the same way as it is done for transaction details.

**Comment by katspaugh** [comment]
I didn't get it – is it for a single tx or for a list? Why is it a POST?

**Comment by rmeissner** [comment]
Do we expect the clients to estimate the `safeTxGas` before? Afaik in the web flow this happens after the decoded data is shown. Actually the same question applies to all other execution related parameters (e.g. `nonce`, etc)

Also do I see it correctly that we don't have a status for the returned tx that would really match?

Could you outline the motivation for choosing the specific input and output objects, compared to other approaches (e.g. having just `to` and `data` as the input and https://safe.global/safe-client-gateway/docs/routes/transactions/models/details/struct.TransactionData.html as the output)?

**Comment by rmeissner** [comment]
What is the `from` field? If it is the Safe, does it really make sense to get information for that one?

**Comment by fmrsabino** [comment]
> Do we expect the clients to estimate the `safeTxGas` before? Afaik in the web flow this happens after the decoded data is shown. Actually the same question applies to all other execution related parameters (e.g. `nonce`, etc)

Regarding the flow this is why I mentioned to align this with web. This should return all the data required for the Review Screen.

> Also do I see it correctly that we don't have a status for the returned tx that would really match?

I didn't understand this question.

> Could you outline the motivation for choosing the specific input and output objects, compared to other approaches (e.g. having just `to` and `data` as the input and https://safe.global/safe-client-gateway/docs/routes/transactions/models/details/struct.TransactionData.html as the output)?

I will check again on this one. The end goal is to have everything that is required for the Review Screen.

**Comment by fmrsabino** [comment]
After the meeting today we can answer some of the following concerns. Metting notes: https://docs.google.com/document/d/1PBV1JLNPY1VbmNf353aiZNgG0Vw_hDlbuTGfS4jpJc8/edit?usp=sharing

> Do we expect the clients to estimate the `safeTxGas` before? Afaik in the web flow this happens after the decoded data is shown. Actually the same question applies to all other execution related parameters (e.g. `nonce`, etc)

We would still expect this from the clients. They would keep the current behaviour in place. This would be an additional call to return the data requested. More details regarding the response/request below.

> Also do I see it correctly that we don't have a status for the returned tx that would really match?

This was raised in the meeting today. More details below.

> Could you outline the motivation for choosing the specific input and output objects, compared to other approaches (e.g. having just `to` and `data` as the input and https://safe.global/safe-client-gateway/docs/routes/transactions/models/details/struct.TransactionData.html as the output)?

The initial motivation was to match the expected flow that the clients had to the `/propose` endpoint thus reducing the number of change required.

However this does not make a lot of sense given that we are dealing with with a transaction that has no status (and it's not even created). We clarified that the scope for this endpoint would be only for transactions that do not exist in any state on our database.

Therefore we can change the request/response with the data that we need to achieve this (still to be defined).

> What is the from field? If it is the Safe, does it really make sense to get information for that one?

We removed this from the scope of this change as the clients prefer the name stored locally either way.

> Should we also do some address decoding on the decoded data? I think this is the case in other client gateway views.

This should be the main nice-to-have – we should provide this if we can map it from the data decoded.

Hopefully clarifies further the proposal (which needs to be updated to reflect the new update).

@rmeissner @DmitryBespalov @MouazAlzahabi @katspaugh @dasanra @johannesmoormann
